### PR TITLE
Parametrize timespan constants.

### PR DESCRIPTION
**Pull request by toxeus**
The constants are parametrized such that libbitcoin can be used by various altcoins by solely editing the config file.

Further constants will be parametrized in upcoming pull requests.

As soon as this patch is in a satisfactory state I'll fix the unit tests and will open PRs in other libbitcoin repos to propagate these changes. Please comment.

### CODE REVIEW

File: src/chain/header.cpp [hunk 1]
```diff
@@ -41,7 +41,7 @@ using wall_clock = std::chrono::system_clock;
 //-----------------------------------------------------------------------------
 
 header::header()
-  : header(0, null_hash, null_hash, 0, 0, 0)
+  : header(0, null_hash, null_hash, 0, 0, 0, settings{})
```
Comment (toxeus) [inline src/chain/header.cpp]: `header::header()` is used by the factory functions below. I added this dummy value such that we can have the discussion here based on the code. I'd fix this by adding a `settings` parameter to this constructor and the factory functions. @evoskuil what do you think?

File: include/bitcoin/bitcoin/chain/chain_state.hpp [hunk 2]
```diff
@@ -124,24 +125,26 @@ class BC_API chain_state
     };
 
     /// Checkpoints must be ordered by height with greatest at back.
-    static map get_map(size_t height, const checkpoints& checkpoints,
+    map get_map(size_t height, const checkpoints& checkpoints,
```
Comment (evoskuil) [inline include/bitcoin/bitcoin/chain/chain_state.hpp]: Keep this static and parameterize necessary settings (or settings object).

File: include/bitcoin/bitcoin/chain/chain_state.hpp [hunk 3]
```diff
@@ -181,29 +184,29 @@ class BC_API chain_state
 
     static activations activation(const data& values, uint32_t forks);
     static uint32_t median_time_past(const data& values, uint32_t forks);
-    static uint32_t work_required(const data& values, uint32_t forks);
+    uint32_t work_required(const data& values, uint32_t forks);
```
Comment (evoskuil) [inline include/bitcoin/bitcoin/chain/chain_state.hpp]: Keep this static and parameterize necessary settings (or settings object).

File: src/chain/header.cpp (hunk 1 above)
Comment (evoskuil) [reply src/chain/header.cpp]: Yes, require the parameter until all other repos have been updated, so that you do not end up missing the necessary settings passage. But at that point you can also add a defaulted construction.

File: src/chain/header.cpp [hunk 4]
```diff
@@ -65,32 +66,35 @@ header::header(const header& other)
...
     nonce_(other.nonce_),
-    metadata(other.metadata)
+    metadata(other.metadata),
+    settings_(other.settings_)
 {
 }
 
 header::header(uint32_t version, hash_digest&& previous_block_hash,
-    hash_digest&& merkle, uint32_t timestamp, uint32_t bits, uint32_t nonce)
+    hash_digest&& merkle, uint32_t timestamp, uint32_t bits, uint32_t nonce, const settings& settings)
```
Comment (evoskuil) [inline src/chain/header.cpp]: Style: honor the 80 character line limit (more above and below).

File: include/bitcoin/bitcoin/settings.hpp [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    uint32_t min_timespan;
+    uint32_t max_timespan;
+
+    // The target number of blocks for 2 weeks of work (2016 blocks).
+    size_t retargeting_interval;
+
+    uint32_t work_limit(bool retarget=true) const
+    {
+        return retarget ? retarget_proof_of_work_limit : no_retarget_proof_of_work_limit;
+    }
```
Comment (evoskuil) [inline include/bitcoin/bitcoin/settings.hpp]: Move implementation to cpp.

File: include/bitcoin/bitcoin/settings.hpp [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+    uint32_t no_retarget_proof_of_work_limit;
+
+    // The upper and lower bounds for the retargeting timespan.
+    uint32_t min_timespan;
+    uint32_t max_timespan;
+
+    // The target number of blocks for 2 weeks of work (2016 blocks).
+    size_t retargeting_interval;
+
+    uint32_t work_limit(bool retarget=true) const
```
Comment (evoskuil) [inline include/bitcoin/bitcoin/settings.hpp]: Style: move declaration above fields.

File: include/bitcoin/bitcoin/settings.hpp [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+    {
+        return retarget ? retarget_proof_of_work_limit : no_retarget_proof_of_work_limit;
+    }
+
```
Comment (evoskuil) [inline include/bitcoin/bitcoin/settings.hpp]: Style: remove blank line.

File: include/bitcoin/bitcoin/settings.hpp [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+#include <cstdint>
+#include <cstddef>
+
+namespace libbitcoin {
+
+/// Common database configuration settings, properties not thread safe.
+class BC_API settings
+{
+
+private:
```
Comment (evoskuil) [inline include/bitcoin/bitcoin/settings.hpp]: Style: private after public: https://github.com/libbitcoin/libbitcoin/wiki/Style-Guide#hpp-format

File: include/bitcoin/bitcoin/settings.hpp [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+
+#include <cstdint>
+#include <cstddef>
+
+namespace libbitcoin {
+
+/// Common database configuration settings, properties not thread safe.
+class BC_API settings
+{
+
```
Comment (evoskuil) [inline include/bitcoin/bitcoin/settings.hpp]: Style: remove blank line: https://github.com/libbitcoin/libbitcoin/wiki/Style-Guide#hpp-format

File: src/chain/header.cpp (hunk 1 above)
Comment (toxeus) [reply src/chain/header.cpp]: Alright, I have propagated the `settings` as a parameter wherever needed. This blew up the PR to some extend. Now, I know what you meant by this refactoring being "pervasive" :) Pushing new version now.

**Comment by toxeus** [comment]
I'm still postponing to fix the tests until I know that the main changes are ok.

File: examples/main.cpp [hunk 10]
```diff
@@ -45,7 +45,8 @@ int bc::main(int argc, char* argv[])
 #endif
 
     // Extracting Satoshi's words from genesis block.
-    const auto block = bc::chain::block::genesis_mainnet();
+    const auto block =
+        bc::chain::block::genesis_mainnet(libbitcoin::settings());
```
Comment (evoskuil) [inline examples/main.cpp]: use `bc::settings()`

Comment (evoskuil) [reply examples/main.cpp]: The genesis block needs to itself become a setting, meaning that `genesis_mainnet` will no longer exist. The genesis block will be deserialized from the base16-encoded setting stream.

File: include/bitcoin/bitcoin/chain/chain_state.hpp [hunk 11]
```diff
@@ -181,29 +184,39 @@ class BC_API chain_state
...
+    static data to_pool(const chain_state& top, const settings& settings);
     static data to_block(const chain_state& pool, const block& block);
-    static data to_header(const chain_state& parent, const header& header);
+    data to_header(const chain_state& parent, const header& header,
+        const settings& settings);
 
-    static uint32_t work_required_retarget(const data& values);
-    static uint32_t retarget_timespan(const chain_state::data& values);
+    static uint32_t work_required_retarget(const data& values,
+        const settings& settings);
```
Comment (evoskuil) [inline include/bitcoin/bitcoin/chain/chain_state.hpp]: Inconsistency here, pass `retarget_proof_of_work_limit`.

File: src/message/compact_block.cpp [hunk 12]
```diff
@@ -99,31 +99,33 @@ bool compact_block::is_valid() const
...
+bool compact_block::from_data(uint32_t version, std::istream& stream,
+    const settings& settings)
 {
     istream_reader source(stream);
-    return from_data(version, source);
+    return from_data(version, source, settings);
 }
 
-bool compact_block::from_data(uint32_t version, reader& source)
+bool compact_block::from_data(uint32_t version, reader& source, const settings& settings)
```
Comment (evoskuil) [inline src/message/compact_block.cpp]: Style: 80 char line length.

**Comment by evoskuil** [comment]
See warnings:
```
  CXX      src/src_libbitcoin_la-settings.lo
src/settings.cpp:33:44: warning: field 'retargeting_factor_' is uninitialized when used here [-Wuninitialized]
    min_timespan(target_timespan_seconds / retargeting_factor_),
                                           ^
src/settings.cpp:34:44: warning: field 'retargeting_factor_' is uninitialized when used here [-Wuninitialized]
    max_timespan(target_timespan_seconds * retargeting_factor_),
                                           ^
src/settings.cpp:35:52: warning: field 'target_spacing_seconds_' is uninitialized when used here [-Wuninitialized]
    retargeting_interval(target_timespan_seconds / target_spacing_seconds_)
                                                   ^
3 warnings generated.
  CXX      src/chain/src_libbitcoin_la-block.lo
  CXX      src/chain/src_libbitcoin_la-chain_state.lo
src/chain/chain_state.cpp:124:29: warning: comparison of unsigned expression >= 0 is always true [-Wtautological-compare]
         (regtest && height >= regtest_bip34_freeze));
                     ~~~~~~ ^  ~~~~~~~~~~~~~~~~~~~~
1 warning generated.
```

File: examples/main.cpp (hunk 10 above)
Comment (toxeus) [reply examples/main.cpp]: Yes, the genesis block is on my todo list. I prefer to approach the migration of the constants iteratively. Once this is merged I'd open a new PR where the genesis block becomes a setting.

**Comment by toxeus** [comment]
@evoskuil I was wondering why I didn't see those warnings. Turns out that they are [enabled for C only](https://github.com/libbitcoin/libbitcoin/blob/52258bf331bb24e3d8e0616cbc1d0fabaad0053a/configure.ac#L307-L332), i.e. they're assigned to `CFLAGS` as opposed to `CXXFLAGS` which is used above and below the snipped I highlighted in the link. Also, `CFLAGS` is not eventually appended to `CXXFLAGS`. Now, I'm wondering why you *did* see the warnings!

File: include/bitcoin/bitcoin/chain/chain_state.hpp (hunk 11 above)
Comment (toxeus) [reply include/bitcoin/bitcoin/chain/chain_state.hpp]: If you check the implementation then you can see that actually four values from `settings` are used in `work_required_retarget()`:
* `settings.retarget_proof_of_work_limit`
* `settings.min_timespan`
* `settings.max_timespan`
* `settings.target_timespan_seconds`.
I'm happy to pass them individually but my judgement was that they would bloat the function signature. Your call.

File: examples/main.cpp (hunk 10 above)
Comment (toxeus) [reply examples/main.cpp]: done

File: src/message/compact_block.cpp (hunk 12 above)
Comment (toxeus) [reply src/message/compact_block.cpp]: done

**Comment by toxeus** [comment]
> See the Travis builds for the warnings.

@evoskuil that was not my point. I've fixed the warnings in the second commit already. I just wanted to raise awareness that `configure.ac` is currently not set up correctly. Maybe @pmienk can confirm that?

**Comment by toxeus** [comment]
gotcha :+1:

**Comment by evoskuil** [comment]
@toxeus The warnings (in fact errors) were the result of field ordering in the header file. Fields are initialized in the header order, not in the order listed in the constructor initializers. I always try to keep these orders the same to avoid this type of failure. The errors were resolved by moving the initialization of these values into the constructor body.

There should not be private fields in the settings class. Making these public and matching the declaration order to the initialization order will resolve the issue properly.

**Comment by toxeus** [comment]
@evoskuil ok, I'm happy to fix the initialization issue the way you've suggested instead of moving it to the constructor body. But what about the build systemt's bug where it doesn't enable warnings for cpp files? Do you prefer me to open an issue for that?

**Comment by evoskuil** [comment]
@toxeus I’m not seeing the issue you are describing. These warnings are in a cpp file. But if you want to open an issue with explanation the libbitcoin-build repository is the best place to do it. Thanks!

**Comment by toxeus** [comment]
@evoskuil I gave you a curl command to see the issue I'm describing. I'll open an issue in the repo you've suggested. Thanks!

**Comment by evoskuil** [comment]
@toxeus I pointed out above the the warnings do appear in the Travis builds. The specific builds from the curl script did not have the warnings only because you inadvertently resolved them.

**Comment by toxeus** [comment]
@evoskuil the curl script fetches the logs for the commit where the warnings should have been present. Visit the travis build link and then click on commit.

**Comment by toxeus** [comment]
@evoskuil Strange. Maybe travis overwrites the logs.

**Comment by evoskuil** [comment]
Just look back at previous builds, the warnings are there in the logs.

**Comment by toxeus** [comment]
@evoskuil wrote:
> Just look back at previous builds, the warnings are there in the logs.

Actually, it's not in the previous build but in the [next](https://travis-ci.org/libbitcoin/libbitcoin/builds/389539885) one. 

Now, I remember how this all came about!
I originally had an initialization order that *did* match the declaration order in the header files in my first version because I also always try to make them match (that's why I couldn't find the warning in the first travis build). Then, you asked me to adhere to the libbitcoin style and move the private members to the end of the header file. I've blindly implemented that suggestion which obviously introduced a mismatch in order between declaration and initialization. And I didn't see the corresponding warnings because I'm using gcc for development and the build system doesn't enable the warnings for gcc. See opened issue [here](https://github.com/libbitcoin/libbitcoin-build/issues/163). So, it turns out that we were both right about the presence of the warnings! :wink:

**Comment by evoskuil** [comment]
Great, thanks for tracking it down. I was only looking at the clang builds. I'll check out the -build issue.

File: include/bitcoin/bitcoin/chain/chain_state.hpp (hunk 11 above)
Comment (evoskuil) [reply include/bitcoin/bitcoin/chain/chain_state.hpp]: Yeah, these aren't always obvious calls. I've tried for a consistent approach whereby classes are constructed with the the settings object given sufficient use, but functions (including static members) are parameterized so as to avoid a dependency on the setting object.

**Comment by evoskuil** [comment]
Time to update the tests and get it passing.

**Comment by toxeus** [comment]
@evoskuil I've fixed the test. Will start work on the PRs for the other repos.

File: src/settings.cpp [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+    target_timespan_seconds(2 * 7 * 24 * 60 * 60),
+    retarget_proof_of_work_limit(0x1d00ffff),
+    no_retarget_proof_of_work_limit(0x207fffff),
+    min_timespan(target_timespan_seconds / retargeting_factor),
+    max_timespan(target_timespan_seconds * retargeting_factor),
+    retargeting_interval(target_timespan_seconds / target_spacing_seconds)
+{
+}
+
+uint32_t settings::work_limit(bool retarget) const
```
Comment (evoskuil) [inline src/settings.cpp]: We should move this function out of settings and into chain::header (see chain::header::proof). Prob should do that in an independent commit.

File: test/chain/block.cpp [hunk 14]
```diff
@@ -262,30 +266,30 @@ BOOST_AUTO_TEST_CASE(block__from_data__insufficient_transaction_bytes__failure)
...
-    chain::block instance;
+    chain::block instance(settings{});
     BOOST_REQUIRE(!instance.from_data(data));
     BOOST_REQUIRE(!instance.is_valid());
 }
 
 BOOST_AUTO_TEST_CASE(block__genesis__mainnet__valid_structure)
 {
-    const auto genesis = bc::chain::block::genesis_mainnet();
+    const auto genesis = bc::chain::block::genesis_mainnet(settings());
```
Comment (evoskuil) [inline test/chain/block.cpp]: Use consistent style (settings() and settings{}).

File: include/bitcoin/bitcoin/settings.hpp [hunk 15]
```diff
@@ -0,0 +1,55 @@
+/**
```
Comment (evoskuil) [inline include/bitcoin/bitcoin/settings.hpp]: Need to add tests for settings class (see other repos).

File: test/chain/block.cpp (hunk 14 above)
Comment (toxeus) [reply test/chain/block.cpp]: My style is very consistent. I always use `settings()` except for cases where the vexing parse occurs.

File: src/settings.cpp (hunk 13 above)
Comment (toxeus) [reply src/settings.cpp]: Independent commit as in independent PR?

**Comment by toxeus** [comment]
Rebased and fixed conflicts.

File: src/settings.cpp (hunk 13 above)
Comment (toxeus) [reply src/settings.cpp]: @evoskuil I've been now looking into moving this function into chain::header as you've suggested. The problem is that it is also used in [chain::chain_state](https://github.com/libbitcoin/libbitcoin/blob/2e1e6792736e73f8a216ea09038084c01f22972a/src/chain/chain_state.cpp#L604) where there's no access to a `header` object. However, there's a comment which states that `bits.self` is unused. If this assignment can be removed then moving `work_limit()` becomes trivial.

Comment (evoskuil) [reply src/settings.cpp]: Yes, I you can set it to zero in that case. Self has no purpose in the context of a pool chain state and when the pool state is upgraded to a block the blocks bits are populated.
